### ISSUE TITLE
Sandboxing portal

### ISSUE THREAD

**Issue by soredake** [flatpak/xdg-desktop-portal#924]
Related:
https://discourse.gnome.org/t/sandboxing-portal/1651
https://github.com/electron/electron/issues/27083
https://github.com/electron/electron/issues/18265
https://github.com/electron/electron/issues/17972#issuecomment-516957971
https://github.com/flatpak/flatpak/issues/3044
https://github.com/flathub/com.valvesoftware.Steam/issues/897

As `CLONE_NEWUSER` and suid is not possible in, it will be nice if there is portal for sandboxing for electron apps/games to use, right now, every electron app or game will not work out of the box in flatpak, apps can use zypak, but games? User will be required to add parameter `--no-sandbox` to game's launch options. If this portal will be added and electron updated to support it apps (that updated to newer electron version) will not need zypak anymore, and at least new electron games will work out of the box.

**Comment by orowith2os** [comment]
Those games you speak of that need the launch argument *should* be able to add the --no-sandbox argument to their launch wrapper until a portal like this is implemented, right? AFAIK it's not possible to flatpak an electron application without a wrapper like that.

**Comment by soredake** [comment]
> Those games you speak of that need the launch argument _should_ be able to add the --no-sandbox argument to their launch wrapper until a portal like this is implemented, right?

Yes, but it's not realistic to assume that all developers will do this. Or users can allow all syscalls (which IIUC will allow `CLONE_NEWUSER`) and electron apps should work out of the box https://github.com/flatpak/flatpak/pull/5224

**Comment by rusty-snake** [comment]
> Or users can allow all syscalls (which IIUC will allow CLONE_NEWUSER) and electron apps should work out of the box https://github.com/flatpak/flatpak/pull/5224

> Depends on ... Use new --disable-userns bubblewrap feature

**Comment by smcv** [comment]
> Or users can allow all syscalls (which IIUC will allow CLONE_NEWUSER)

No, Flatpak cannot ever allow sandboxed apps to use `CLONE_NEWUSER`: that would break Flatpak's security model.

**Comment by smcv** [comment]
> portal for sandboxing

This is not really an actionable request: there is an unlimited number of things that it could mean.

For Electron specifically, there is already a portal for running things in a more restrictive Flatpak sandbox ("sub-sandboxing" as implemented by the `flatpak-portal` service), and that's how flatpak-spawn and Zypak work; but it is part of Flatpak, not part of xdg-desktop-portal, so it will not work for Snap (or for any other sandboxing frameworks that are alternatives to Flatpak and Snap). I think that makes sense: xdg-desktop-portal doesn't know the finer details of how a Flatpak app works, but Flatpak does. Similarly, if Electron apps want to create a sub-sandbox in Snap, it would make more sense to ask Snap to set that up, because xdg-desktop-portal doesn't know the finer details of how a Snap app works, but Snap does.

> I'm also interested in if having a portal like this would make it easier to make sub-sandboxed environments, such as a package flatpakking podman and running it in that flatpak environment

Sorry, this is not going to be implementable without allowing a full sandbox escape. podman can only do what it does by using a setuid-root executable, which is scary attack surface and cannot work inside a Flatpak sandbox.

Flatpak reduces its attack surface by limiting itself to things that can work with a non-setuid-root bubblewrap on modern kernels, but the price we pay for that is that we can only have two user IDs ("me" and "not me") and no setuid, and that's not enough for what podman wants.

For Flatpak apps that have the necessary permissions to do an unrestricted sandbox escape, it would be possible to run podman that way, but then that app is just as trusted as if it was not sandboxed at all (it can run arbitrary code with the user's privileges).
